operator/rediscontroller/controller_test.go

```diff
@@ -0,0 +1,296 @@
+//go:build integration
+
+package rediscontroller
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+	"testing"
+	"time"
+
+	exoscalev1 "github.com/vshn/provider-exoscale/apis/exoscale/v1"
+	"github.com/vshn/provider-exoscale/internal/operatortest"
+	"github.com/vshn/provider-exoscale/operator/mapper"
+
+	xpv1 "github.com/crossplane/crossplane-runtime/apis/common/v1"
+	"github.com/crossplane/crossplane-runtime/pkg/event"
+	exoscaleapi "github.com/exoscale/egoscale/v2/api"
+	"github.com/exoscale/egoscale/v2/oapi"
+	"github.com/stretchr/testify/mock"
+	"github.com/stretchr/testify/suite"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/utils/pointer"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type RedisControllerTestSuite struct {
+	operatortest.Suite
+
+	timeout  time.Duration
+	interval time.Duration
+
+	p *pipeline
+}
+
+func (ts *RedisControllerTestSuite) SetupTest() {
+	ts.timeout = 10 * time.Second
+	ts.interval = 250 * time.Millisecond
+
+	nopRecorder := event.NewNopRecorder()
+	ts.p = newPipeline(ts.Client, nopRecorder, ts.ExoClientMock)
+
+	err := SetupControllerWithConnecter(ts.Manager, "testredis", nopRecorder, &connector{
+		kube:     ts.Client,
+		recorder: nopRecorder,
+		p:        ts.p,
+	}, 0)
+	ts.Require().NoError(err)
+}
+
+func (ts *RedisControllerTestSuite) TestHappyPath() {
+	settings := map[string]any{
+		"maxmemory_policy": "noeviction",
+	}
+	mg := newRedisInstance("test", settings)
```

Comment:
This seems like a rather large test case that tests everything :tm: . This makes maintaining the test very hard IMO.
Is there no chance to test individual actions in isolation of eachother? Like
- Test if creation works: start with nothing, end up with a "running" instance
- Test if update works: start with an "existing" instance, update a field, await status updates
- Test if delete works: start with "existing" instance, delete it. assert nothing exists.

To me, create, update and delete are three distinct operations worth their own test case. In case "creation" of an instance fails, it shouldn't fail the update test case as well, for example. Or deletion.

Reply:
Yeah, that does make sense. However, for the non-creation cases, I'd need to still do the creation regardless since I (correct me if I'm wrong) can't add objects just like that in the envtest environment. 

I can certainly add this, but not sure how much it improves the code TBH.

Reply:
> can't add objects just like that in the envtest environment.

sure you can, you just need to prepare the spec and update status accordingly, so that the reconciler thinks it's already created from a previous "loop".